connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSourceTask.java

```diff
@@ -370,9 +370,12 @@ private RecordHeaders convertHeaderFor(SourceRecord record) {
         return result;
     }
 
-    private void commitTaskRecord(SourceRecord record) {
+    private void commitTaskRecord(SourceRecord record, RecordMetadata metadata) {
         try {
             task.commitRecord(record);
+            if (metadata != null) {
+                task.commitRecord(record, metadata);
```

Comment:
Hmm, why do we need to call commitRecord twice?

Reply:
A SourceTask implementation is unlikely to implement both methods. They are nops by default.

If it does, it's unlikely an implementation would do the same thing in both methods.

If it does, the interface does not guarantee that commitRecord() is called exactly once per record anyway (or even at least once). And indeed other methods, e.g. stop() and commit(), make no such guarantees either.

I'd be in favor of deprecating commitRecord(record) to avoid confusion here, but that is out of scope for this KIP and PR.

Reply:
Since this is a public API change, I think it's important to think through the impact to  connector developers. I am thinking that we can (1) provide a default implementation of commitRecord(SourceRecord record, RecordMetadata metadata) in SourceTask that calls commitRecord(SourceRecord record) by ignoring RecordMetadata to provide backward compatibility; (2) only use commitRecord(SourceRecord record, RecordMetadata metadata) in WorkerSourceTask; (3) mark commitRecord(SourceRecord record) as deprecated to encourage people to use the new API and document the behavior if both methods are implemented (the implementation for the old api will be ignored). @ewencp @rhauch Any thought on this?

Reply:
So far, the Connect API exposes only a few Kafka client types in a few areas (right now just `SinkTask`). We should be careful and explicit about making the `SourceTask` interface use `RecordMetadata`.

But assuming that's okay, I think having Connect call two distinct `commitRecord(...)` methods with different signatures is at best confusing for developers (which do I implement) and at worst a potential source of error. We have resolved similar evolutions in the past by following the approach that @junrao mentioned, and I think that makes the most sense here. (We've not always deprecated the older method when it still applicable, but in this case I think we'd want to deprecate the older method.)

However, before we introduce a new variant of an existing method, we should also consider whether we might need to again modify the signature in the future. If so, we should consider whether it makes sense to create a new interface type to pass to the method that would make it easier in the future. I'm not sure that we do, since the `WorkerSourceTask` calls this from within the producer callback and only when the exception is not null, meaning there currently is no other data available to the invocation.

Reply:
Happy to deprecate the older method if we have concensus here.

Reply:
given we can use default implementations now, just providing a default implementation and not deprecating seems better. afaik, `RecordMetadata` has never been requested before, but a number of connectors use the existing signature for `commitRecord`. Calling twice and having to understand that does seem confusing, but having overloads and simplified versions doesn't -- you just get to choose to implement something simpler. We've been pretty conservative with deprecation in the connect and that has served us well wrt broad compatibility -- I think only `onPartitionsAssigned/Revoked` for sink tasks and a couple of configs have gone through this, and still haven't been removed (with little maintenance cost afaik), which means we've remained cleanly compatible in most ways all the way back to 0.9. I'm generally for deprecation and removal (e.g. there's lots of core configs that I think are unnecessary and could be cleaned up), but only if the cost is worth it. In this case, there *is* a transition cost since the existing API is being used, and I'm not sure the reduction in confusion is worth it (vs some simple API docs explaining default impl).

Reply:
Good point about not deprecating, @ewencp. I could definitely go either way, and definitely see value in not deprecating and leaving it up to the implementer to choose which they'd implement. But I do think overriding one method is better than having to potentially override both methods for different purposes.

Because this is changing the Connect API, it does seem like we need to come to consensus and approve KIP-416 before this PR can be merged, though.

Reply:
My vote is for keeping both methods (as implemented here), since this doesn't break or deprecate existing code, nor is it especially confusing to see the same method with overloaded parameters in an interface, especially when neither are required to implement. I think the confusing part is just in the WorkerSourceTask implementation, not the SourceTask interface -- it's a little surprising to see the same method called twice here. I suggest we add a comment here and move on.

@AndrewJSchofield thoughts?

Reply:
I'd prefer to see KIP-416 introduce `public void commitRecord(SourceRecord sourceRecord, RecordMetadata recordMetadata)` which is called in exactly the same situations as the existing `public void commitRecord(SourceRecord sourceRecord)`. The new method might be called with null record metadata. The implementor has a choice of which to implement.

Reply:
@ryannedolan, I disagree that the confusing part is just in the `WorkerSourceTask` implementation. I believe it is confusing that both methods might be called with the same record, and that there is no benefit from doing that. So the issue I have is that I believe the proposed changes to the Connect API make it less clear how to properly implement it.

It is much clearer for the API to offer two methods and to say the framework always calls the new one, the new one by default calls the old one, and that task implementations can override one or none of them. Then it is entirely up to the task implementation to decide whether to even treat these conditions as distinct.

Reply:
@rhauch just clarifying, are you suggesting we take @junrao 's suggestion, except we don't deprecate the existing method? i.e. have the new method call the old?

@AndrewJSchofield does that satisfy your concern as well?

Reply:
@ryannedolan, yes, I am suggesting that we follow @junrao's suggestion minus deprecating the old method. IOW, we'd do the following:
1) add the new method that by default calls the old method; and
2) change all uses in the Connect runtime (namely in WorkerSinkTask and its test class) to use the new method

and update KIP-416 accordingly.

Reply:
Sounds like we have concensus. Will update shortly.

Reply:
I made the change. Happy with either approach.